src/query/src/range_select/plan_rewrite.rs

```diff
@@ -0,0 +1,424 @@
+                                index += 1;
+                                continue;
+                            }
+                            if range_fn.fill.is_empty() {
+                                range_fn.fill = str.clone();
+                                index += 1;
+                                continue;
+                            }
+                            let align =
+                                parse_duration(str.as_str()).map_err(DataFusionError::Internal)?;
+                            if self.align != Duration::default() && self.align != align {
+                                return Err(DataFusionError::Internal(
+                                    "Inconsistent align given in Range Function Rewrite".into(),
+                                ));
+                            } else {
+                                self.align = align
+                            }
+                        }
+                        _ => {
+                            return Err(DataFusionError::Internal(
+                                "Unexpected param in range_fn args".into(),
+                            ))
+                        }
+                    }
+                    index += 1
+                }
+                if !self.by.is_empty() && self.by != by {
+                    return Err(DataFusionError::Internal(
+                        "Inconsistent by given in Range Function Rewrite".into(),
+                    ));
+                } else {
+                    self.by = by.clone();
+                }
+                range_fn.expr = self.gen_range_expr(func_name, args)?;
+                let alias = Expr::Column(Column::from_name(range_fn.expr.display_name()?));
+                self.range_fn.push(range_fn);
+                return Ok(alias);
+            }
+        }
+        Ok(node)
+    }
+}
+
+pub struct RangePlanRewriter {
+    table_provider: DfTableSourceProvider,
+    context_provider: DfContextProviderAdapter,
+}
+
+impl RangePlanRewriter {
+    pub fn new(
+        table_provider: DfTableSourceProvider,
+        context_provider: DfContextProviderAdapter,
+    ) -> Self {
+        Self {
+            table_provider,
+            context_provider,
+        }
+    }
+
+    pub async fn rewrite(&mut self, plan: LogicalPlan) -> Result<LogicalPlan> {
+        match self.rewrite_logical_plan(&plan).await? {
+            Some(new_plan) => Ok(new_plan),
+            None => Ok(plan),
+        }
+    }
+
+    #[async_recursion]
+    async fn rewrite_logical_plan(&mut self, plan: &LogicalPlan) -> Result<Option<LogicalPlan>> {
+        let inputs = plan.inputs();
+        let mut new_inputs = Vec::with_capacity(inputs.len());
+        for input in &inputs {
+            new_inputs.push(self.rewrite_logical_plan(input).await?)
+        }
+        match plan {
+            LogicalPlan::Projection(Projection { expr, input, .. })
+                if have_range_in_exprs(expr) =>
```

Comment:
we should also check if all expressions are range fn. we cannot handle a projection containing normal expression and range fn simultaneously

Reply:
The rule is the same as aggregator plan: only group col and aggregator are allowed to present. If read a query like `SELECT val_0, avg(val_0) range '5m'` or `SELECT tag_0, avg(val_0) range '5m' by (tag_1)` we should report an error

Reply:
that's right, I will add some additional check to make sure only group col 、aggregator、time index column will show in projection. 
To be honest, if you really write other columns, it will fail during the execution process, because the range plan does not provide the schema info of these columns, and the projection cannot find these columns, but theoretically these problems should be solved in the logical plan stage

Reply:
In fact we don't need to check situation like `SELECT val_0, avg(val_0) range '5m'`, our addtional projection plan will check projection expr base on our range plan schema,   `val_0` not exsit in our range plan schema, so the query will failed on logical plan stage

Reply:
No, we'd better do our best to validate. You should ensure the plan generated by your rule is legal so other stages can rely on their input. And we won't always have an additional projection, like discussed in another comment